builtin/logical/pki/crl_util.go

```diff
@@ -0,0 +1,153 @@
+	}
+	if cert == nil {
+		return nil, fmt.Errorf("Got a nil certificate")
+	}
+
+	if cert.NotAfter.Before(time.Now()) {
+		return nil, nil
+	}
+
+	revInfo := revocationInfo{
+		CertificateBytes: certEntry.Value,
+		RevocationTime:   time.Now().Unix(),
+	}
+
+	certEntry, err = logical.StorageEntryJSON("revoked/"+serial, revInfo)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating revocation entry")
+	}
+
+	err = req.Storage.Put(certEntry)
+	if err != nil {
+		return nil, fmt.Errorf("Error saving revoked certificate to new location")
+	}
+
+	err = buildCRL(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error encountered during CRL building: %s", err)
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"revocation_time": revInfo.RevocationTime,
+		},
+	}, nil
+}
+
+func buildCRL(req *logical.Request) error {
+	revokedSerials, err := req.Storage.List("revoked/")
+	if err != nil {
+		return fmt.Errorf("Error fetching list of revoked certs: %s", err)
+	}
+
+	revokedCerts := []pkix.RevokedCertificate{}
+	var revInfo revocationInfo
+	for _, serial := range revokedSerials {
+		revokedEntry, err := req.Storage.Get("revoked/" + serial)
+		if err != nil {
+			return fmt.Errorf("Unable to fetch revoked cert with serial %s: %s", serial, err)
+		}
+		if revokedEntry == nil {
+			return fmt.Errorf("Revoked certificate entry for serial %s is nil", serial)
+		}
+		if revokedEntry.Value == nil || len(revokedEntry.Value) == 0 {
+			// TODO: In this case, remove it and continue? How likely is this to
+			// happen? Alternately, could skip it entirely, or could implement a
+			// delete function so that there is a way to remove these
+			return fmt.Errorf("Found revoked serial but actual certificate is empty")
+		}
+
+		err = revokedEntry.DecodeJSON(&revInfo)
+		if err != nil {
+			return fmt.Errorf("Error decoding revocation entry for serial %s: %s", serial, err)
+		}
+
+		revokedCert, err := x509.ParseCertificate(revInfo.CertificateBytes)
+		if err != nil {
+			return fmt.Errorf("Unable to parse stored revoked certificate with serial %s: %s", serial, err)
+		}
+
+		if revokedCert.NotAfter.Before(time.Now()) {
+			err = req.Storage.Delete(serial)
+			if err != nil {
+				return fmt.Errorf("Unable to delete revoked, expired certificate with serial %s: %s", serial, err)
+			}
+			continue
+		}
+
+		revokedCerts = append(revokedCerts, pkix.RevokedCertificate{
+			SerialNumber:   revokedCert.SerialNumber,
+			RevocationTime: time.Unix(revInfo.RevocationTime, 0),
+		})
+	}
+
+	rawSigningBundle, caCert, err := fetchCAInfo(req)
+	if err != nil {
+		return fmt.Errorf("Could not fetch the CA certificate")
+	}
+
+	signingPrivKey, err := rawSigningBundle.getSigner()
+	if err != nil {
+		return fmt.Errorf("Unable to get signing private key: %s", err)
+	}
+
+	// TODO: Make expiry configurable
+	crlBytes, err := caCert.CreateCRL(rand.Reader, signingPrivKey, revokedCerts, time.Now(), time.Now().Add(time.Hour*72))
```

Comment:
I'll make it a var, so that it can easily be overridden if there is a different configured lifetime.

It's not just a shorter expiration that someone might want -- they may also want a longer expiration. It really depends on how often they are going to hit the CRL, what caching behavior they want on the clients, etc.

I think 3 days is a reasonable default, and I didn't think it was worth making it configurable at this moment in time, but I can see organizations wanting control over it. People have very particular certificate policies.